packages/path_provider/path_provider_platform_interface/lib/src/method_channel_path_provider.dart

```diff
@@ -0,0 +1,90 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'enums.dart';
+
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
+import 'package:path_provider_platform_interface/path_provider_platform_interface.dart';
+import 'package:platform/platform.dart';
+
+/// An implementation of [PathProviderPlatform] that uses method channels.
+class MethodChannelPathProvider extends PathProviderPlatform {
+  /// The method channel used to interact with the native platform.
+  @visibleForTesting
+  MethodChannel methodChannel =
+      MethodChannel('plugins.flutter.io/path_provider');
+
+  // Ideally, this property shouldn't exist, and each platform should
+  // just implement the supported methods. Once all the platforms are
+  // federated, this property should be removed.
+  Platform _platform = const LocalPlatform();
+
+  /// This API is only exposed for the unit tests. It should not be used by
+  /// any code outside of the plugin itself.
+  @visibleForTesting
+  void setMockPathProviderPlatform(Platform platform) {
+    _platform = platform;
+  }
+
+  Future<String> getTemporaryPath() async {
+    return methodChannel.invokeMethod<String>('getTemporaryDirectory');
+  }
+
+  Future<String> getApplicationSupportPath() async {
+    return methodChannel.invokeMethod<String>('getApplicationSupportDirectory');
+  }
+
+  Future<String> getLibraryPath() async {
+    if (!_platform.isIOS && !_platform.isMacOS) {
+      throw UnsupportedError('Functionality only available on iOS/macOS');
+    }
+    return methodChannel.invokeMethod<String>('getLibraryDirectory');
+  }
+
+  Future<String> getApplicationDocumentsPath() async {
+    return methodChannel
+        .invokeMethod<String>('getApplicationDocumentsDirectory');
+  }
+
+  Future<String> getExternalStoragePath() async {
+    if (!_platform.isAndroid) {
+      throw UnsupportedError('Functionality only available on Android');
+    }
+    return methodChannel.invokeMethod<String>('getStorageDirectory');
+  }
+
+  Future<List<String>> getExternalCachePaths() async {
+    if (!_platform.isAndroid) {
+      throw UnsupportedError('Functionality only available on Android');
+    }
+    final List<String> paths = await methodChannel
+        .invokeListMethod<String>('getExternalCacheDirectories');
```

Comment:
You don't need to await here to apply the transformation, you can just return the future returned by doing `.then` on the result of `invokeListMethod`:

```
return methodChannel
    .invokeListMethod<String>('blah')
    .then((List<String> paths) => paths.map((String p) => p).toList());
```

The user of this API needs to await for the `Future` that you're returning here, so there's no need for this method to await. You can just return something that will eventually generate a `List<String>` and be done quickly.

Then, this method doesn't need to be `async` anymore ;)

Reply:
See response above :)